### PR TITLE
Add improved top entity logic

### PR DESCRIPTION
**Pull request by kschiffer**
#### Summary

This PR adds a highly extended top entities functionality to the Console redesign. The Console determine "top entities" by combining the user-set bookmarks with a recency/frequency system, which ranks entities based on their recency and frequency of use. The top entities are used in the sidebar, in the search panel (when no query is entered yet) and in the top entities panel in the general overview.

image
image
image
image

#### Changes

- Add new top entities store logic
- Update frequency recency lib
- Add the top entities to sidebar, search panel and top entities panel
- Change event logic to allow for multiple streams for the same entity type being open concurrently
- Move tracking entity access from entity fetching logic to entity views
- Add caching mechanism to some fetch logics
- Various related small changes

##### Steps

- Add bookmarks and visit entities in the Console
- See these entities appear in the sidebar/search panel and top entities panel
- Observe the sort order changing based on how often an entity is accessed

#### Notes for Reviewers

The diff is quite large, but I've tried to inline-document everything as good as possible. Please ask me if anything is unclear.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Testing: The steps/process to test this feature are clearly explained including testing for regressions.
- [ ] Infrastructure: If infrastructural changes (e.g., new RPC, configuration) are needed, a separate issue is created in the infrastructural repositories.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/webui/console/containers/top-entities-dashboard-panel/all-top-entities/index.js [hunk 1]
```diff
@@ -25,37 +27,51 @@ import LastSeen from '@console/components/last-seen'
...
+      getValue: entity => entity,
+      render: ({ entity, id }) =>
+        Boolean(entity?.name) ? (
+          <>
+            <Message content={entity.name} component="p" className="mt-0 mb-cs-xs p-0 fw-bold" />
+            <Message content={id} component="span" className="c-text-neutral-light" />
+          </>
+        ) : (
+          <Message content={id} component="p" className="mt-0 mb-cs-xs p-0 fw-bold" />
+        ),
```
Comment (ryaplots) [inline pkg/webui/console/containers/top-entities-dashboard-panel/all-top-entities/index.js]: Why using the `<Message />` component here without an intl message? I had the idea we should use plan tags in these cases.

File: pkg/webui/console/containers/top-entities-dashboard-panel/top-applications/index.js [hunk 2]
```diff
@@ -40,32 +37,32 @@ const m = defineMessages({
...
-      ),
+      getValue: entity => entity,
+      render: ({ entity: { name }, id }) =>
+        Boolean(name) ? (
+          <>
+            <Message content={name} component="p" className="mt-0 mb-cs-xs p-0 fw-bold" />
+            <Message content={id} component="span" className="c-text-neutral-light" />
+          </>
+        ) : (
+          <Message content={id} component="p" className="mt-0 mb-cs-xs p-0 fw-bold" />
```
Comment (ryaplots) [inline pkg/webui/console/containers/top-entities-dashboard-panel/top-applications/index.js]: same comment as above

File: pkg/webui/console/containers/top-entities-dashboard-panel/top-devices/index.js [hunk 3]
```diff
@@ -37,23 +34,26 @@ const m = defineMessages({
...
-      ),
+      getValue: entity => entity,
+      render: ({ entity: { name }, id }) =>
+        Boolean(name) ? (
+          <>
+            <Message content={name} component="p" className="mt-0 mb-cs-xs p-0 fw-bold" />
+            <Message content={id} component="span" className="c-text-neutral-light" />
+          </>
+        ) : (
+          <Message content={id} component="p" className="mt-0 mb-cs-xs p-0 fw-bold" />
```
Comment (ryaplots) [inline pkg/webui/console/containers/top-entities-dashboard-panel/top-devices/index.js]: same as above

File: pkg/webui/console/containers/top-entities-dashboard-panel/top-gateways/index.js [hunk 4]
```diff
@@ -36,38 +33,41 @@ const m = defineMessages({
...
-      ),
+      getValue: entity => entity,
+      render: ({ entity: { name }, id }) =>
+        Boolean(name) ? (
+          <>
+            <Message content={name} component="p" className="mt-0 mb-cs-xs p-0 fw-bold" />
+            <Message content={id} component="span" className="c-text-neutral-light" />
+          </>
+        ) : (
+          <Message content={id} component="p" className="mt-0 mb-cs-xs p-0 fw-bold" />
```
Comment (ryaplots) [inline pkg/webui/console/containers/top-entities-dashboard-panel/top-gateways/index.js]: same as above

**Review by ryaplots — CHANGES_REQUESTED** [review]
This is really awesome. Great work! 🎉  See my small comment on the usage of message and please fix the application `lastSeen` in the general view of the top entities.

File: pkg/webui/console/containers/top-entities-dashboard-panel/all-top-entities/index.js (hunk 1 above)
Comment (kschiffer) [reply pkg/webui/console/containers/top-entities-dashboard-panel/all-top-entities/index.js]: Very right. I think that's an oversight from refactoring the old version of the top entities panel.

**Comment by kschiffer** [comment]
I had to substantially refactor the implementation to address a couple of issues:

*   Top entities only appearing after a refresh
*   Same top entities being applied to different logged in users
*   Top entities causing issues when the entity has been deleted

Instead of independently storing the recency frequency items into local storage, the implementation now works with a redux middleware which stores the data into localstorage and automatically keeps it updated upon changes. Also, the stored data is now keyed by the user id.

Since the top entities can be entirely derived from the bookmarks and recency frequency items, the items are not stored into the store anymore but composed on the spot by the top entities selector. The top entity logic now only takes care of fetching required entities that are not already in the store.

Please do another review round with these new changes :pray:

File: pkg/webui/console/containers/top-entities-dashboard-panel/item.js [hunk 5]
```diff
@@ -12,51 +12,59 @@
...
+      // Only start the stream if the entity is already fetched
+      // so we avoid starting streams for entities that don't exist
+      // or we don't have permissions for.
+      return
+    }
+    if (type === END_DEVICE) {
+      dispatch(startDeviceEventsStream(id, { filter: [EVENT_END_DEVICE_HEARTBEAT_FILTERS_STRING] }))
+    } else if (type === APPLICATION) {
+      dispatch(
+        startApplicationEventsStream(id, { filter: [EVENT_END_DEVICE_HEARTBEAT_FILTERS_STRING] }),
```
Comment (ryaplots) [inline pkg/webui/console/containers/top-entities-dashboard-panel/item.js:49]: Although the application events are being fetched here, they are not being updated in the panel:
Screenshot 2024-07-17 at 15 43 33

File: pkg/webui/console/containers/sidebar/navigation/top-entities-section.js [hunk 6]
```diff
@@ -13,68 +13,110 @@
...
-        />
-      )}
+      <SectionLabel label={topEntities.length === 0 ? noneLabel : label} icon={IconPlus} />
+      <RequireRequest
+        requestAction={getTopEntities()}
+        spinnerProps={{ inline: true, micro: true, center: true, className: 'mt-ls-s' }}
+        errorRenderFunction={SectionError}
+      >
+        {topEntities.slice(0, 6).map((topEntity, index) => (
+          <TopEntityItem key={index} {...topEntity} />
```
Comment (ryaplots) [inline pkg/webui/console/containers/sidebar/navigation/top-entities-section.js:95]: Clicking on a top end device on the side bar in the end device overview does not lead anywhere, although the path changes. The page does not re-render.

File: pkg/webui/console/views/overview/index.js [hunk 7]
```diff
@@ -17,45 +17,31 @@ import React from 'react'
...
-          <NotificationsDashboardPanel />
-        </div>
-        <div className="item-12 lg-xl:item-12 xl:item-6 md-lg:item-6">
-          <DocumentationDashboardPanel />
-        </div>
-        <div className="item-12 lg-xl:item-12 xl:item-6 md-lg:item-6">
-          <ShortcutPanel />
-        </div>
+    <div className="container container--xl grid p-ls-s gap-ls-s md:p-cs-xs md:gap-cs-xs">
+      <div className="item-12 md:item-12 lg:item-6 sm:item-6">
```
Comment (ryaplots) [inline pkg/webui/console/views/overview/index.js]: Here the responsiveness was overwritten, the classes should be: 
```suggestion
      <div className="item-12 lg-xl:item-12 xl:item-6 md-lg:item-6">
```

**Review by ryaplots — CHANGES_REQUESTED** [review]

**Review by ryaplots — APPROVED** [review]

**Pull request merged**